### ISSUE TITLE
Submit objects without Activity

### ISSUE THREAD

**Issue by xray7224** [w3c-social/activitypub#32]
It's been suggested to simply posting new content we allow posting of just the object to the feed and as a side effect have an activity generated. This came up as part of the convergence effort with micropub, they want to make it as simple as possible to submit new objects. It would seem like there is no information on an Activity that could not be generated by the server for posting, we already generate activities when people do a DELETE or PUT on the object ID.

**Comment by rhiaro** [comment]
+1

And if the activity generated is effectively identical to the object being created, then we could forego the activity altogether and pass the object around directly. (Especially since a lot of example activities seem to have the whole object embedded in them anyway, and the activity doesn't necessarily add anything but an extra layer of abstraction)

**Comment by cwebber** [comment]
I'm sympathetic to this in a sense, and @oshepherd has argued for this as well.  I appreciate how it would help bring things closer together between the specs.

However, we had a conversation about this at the W3C Face to Face, and @evanp made very strong arguments against it.  His rationale was that nothing else in the feed is object-only.  I suggested that maybe posting client to server just the object could imply a Create, but Evan felt that this was the only activity which was implied, so it wasn't helpful to make an exception for it, which is a reasonable point.

I'm inclined to stick with Evan on this one because he has the most experience here and to close this, but I'll leave a bit of opportunity for counter-arguments.

**Comment by oshepherd** [comment]
The activity-object dichotomy isn't really a problem for a server - if anything it might make the server cleaner. But my experience is that it is perhaps *the greatest* issue with the pump.io protocol that "post" objects exist. Why? Because, among other things, "post" objects carry the permission information.

I can't, among other things, know the audience of an object without seeing its' post activity. The post activity is entirely superfluous - it exists purely to carry that information. But that information is a property of the object! It is really quite strange that the permissions applying to an object are carried by an independent activity object.

You can see the complexity this causes in the pump.io codebase by looking for all the places it has to find the post activity for a given object. Some of the weirdest server corner cases come when you manage to have an object with no corresponding post (this was a cause of much inexplicable server weirness in the development of Impeller). It really causes problems for clients where you want to Cc: somebody on a reply to an object or whatever and the client has no idea what the original destinations were.

Actually, I think the best argument against the "post" activity is that pump.io divides things into major and minor activities. A major activity is defined by `(a.verb == 'post' && a.object.objectType != 'comment') || a.verb == 'share'`.

I have problems with both major activities. Post, because its' content is vacuous, and share, because what I actually want is a 'share obejct' - something which I can attach some commentary to.

In other words, I'd reduce the only major activity down to 'post', and at that point you have to ask why it exists at all? It is false orthogonality in my opinion - it is false because that one activity is special, and that one special activity serves no purpose.

We should not have a 'feed of activities' - what we should have is a 'feed of objects, some of which are activities'.

**Comment by xray7224** [comment]
I see both sides of this and honestly, I'm not sure which side I fall down on. The argument from Evan is a good one, everything else you post to the feed is an activity and to make it so you post an object only introduces an exception to the rule, this makes it conceptually way more complicated, I would also argue server side the code to process activities probably would be more complicated as it then has to deal with the fact both activities and objects can be posted but if it gets an object it infers a create activity.

I think removing the create activity all together causes us some problems. The feed currently serves as two purposes: a log of what has happened and comments (instructions to servers to do certain things). This would in my opinion probably break that as suddenly there is no create activity. The purpose for the create activity in this case is to outlive the object and be able to describe the creation of an activity.

In general I'd argue the audience targeting isn't too bad as internal lists should be kept for who the object has been sent to, this covers a case when a object has been sent to a user in a group but then later the user has been removed from the group. Trying to find list of the audience for an object can be achieved that way.

I personally find the idea of having a "create" activity very useful, I even find using one to create objects very useful. I am of the opinion having the "create" activity simplifies the specification and is much easier for the client side developers. I however remain on the fence about allowing people to post just the object and having the server create the create activity as convergence with micropub is very important and it won't change anything fundamental if it's optional (i.e. if posting a create activity is still accepted).

**Comment by oshepherd** [comment]
> I see both sides of this and honestly, I'm not sure which side I fall down on. The argument from Evan is a good one, everything else you post to the feed is an activity and to make it so you post an object only introduces an exception to the rule, this makes it conceptually way more complicated, I would also argue server side the code to process activities probably would be more complicated as it then has to deal with the fact both activities and objects can be posted but if it gets an object it infers a create activity.
> 
> I think removing the create activity all together causes us some problems. The feed currently serves as two purposes: a log of what has happened and comments (instructions to servers to do certain things). This would in my opinion probably break that as suddenly there is no create activity. The purpose for the create activity in this case is to outlive the object and be able to describe the creation of an activity.

Removing the "create" activity makes your feed a feed of objects, not activities. Activities, after all, are objects. No isomorphism is broken here - it just becomes slightly different.

If it is necessary to preserve information on the creation of a deleted object, then the traditional method is to convert the object into a "tombstone" - ideally the server would already do this anyway such that it can return "410 Gone" errors for accesses to the object. Maybe AS2 needs a "tombstone" object type ("Deleted" may be a better name here)

> In general I'd argue the audience targeting isn't too bad as internal lists should be kept for who the object has been sent to, this covers a case when a object has been sent to a user in a group but then later the user has been removed from the group. Trying to find list of the audience for an object can be achieved that way.

Internal lists are fine for the server. How are they surfaced to *clients*? This has been a very common complaint from pump.io client developers

> I personally find the idea of having a "create" activity very useful, I even find using one to create objects very useful. I am of the opinion having the "create" activity simplifies the specification and is much easier for the client side developers. I however remain on the fence about allowing people to post just the object and having the server create the create activity as convergence with micropub is very important and it won't change anything fundamental if it's optional (i.e. if posting a create activity is still accepted).

I don't see any simplification, actually complexity (in the need to specify create activities). Most of this goes away if you just permit the feed to contain objects - and note the semantics of that.

ActivityStreams is the odd one out in the social world for having a "post"/"create" activity.

**Comment by cwebber** [comment]
Okay, glad to hear some good arguments being made here.  I'm thinking about this more and starting to realize that I'm even more on the fence than I was.

> Internal lists are fine for the server. How are they surfaced to clients? This has been a very common complaint from pump.io client developers

That's a really good point.

> ActivityStreams is the odd one out in the social world for having a "post"/"create" activity.

Even with that, I think our usage of it in ActivityPump isn't the way it was intended, at least for client to server.  @jasnell said it was not the intention to use "Create" as a verb from client to server to initiate creating an object, but that it makes sense to wrap something in Create in the "public log" of activities, because then you're *showing* that something was created.

This direction would make a lot of sense, because yes, from a server to server point of view you're expressing that something was created, but the ACL attached to the object would be the *wrapped* object's ACL, not the wrapper.

**Comment by cwebber** [comment]
@oshepherd How would you feel about this:

* A user from client to server can post a non-activity object (something with no activity-type side effects, and does not inherit from Activity), and that counts as creating the thing.  That object is also the one that specifies its ACL.
* When distributed in a server to server fashion, or viewed in a public log, that object is wrapped in a Create.  However, the ACL attached to the object is not the ACL of the Create, but the ACL of the internal object.  (The ACL of the Create is just the ACL of who can see that something was created, which is *probably* copied from the internal object, but in the "long run" of the object, does not affect things much.)

**Comment by cwebber** [comment]
We agreed to add for the _client to server api only_ that clients can post non-activity objects and they will become wrapped in Create activities with the addressing (to/cc/bcc/bto) moved onto the Create object.

This is a major step towards convergence.

**Comment by tsyesika** [comment]
I've pushed a fix for this :)
